TSPL.docc/ReferenceManual/Attributes.md

```diff
@@ -210,8 +210,76 @@ including important milestones.
   obsoleted: <#version number#>
   ```
   The *version number* consists of one to three positive integers, separated by periods.
+
+- The `noasync` argument indicates that the declared symbol can't be used directly
+  in an asynchronous context.
+
+  Because Swift concurrency can resume on a different thread
+  after a potential suspension point,
+  using elements like thread-local storage, locks, mutexes, or semaphores
+  across suspension points can lead to incorrect results.
+
+  To avoid any errors, add an `@available(*, noasync)` attribute to the symbol's declaration:
+
+  ```swift
+  extension pthread_mutex_t {
+    
+    @available(*, noasync)
+    mutating func lock() {
+        pthread_mutex_lock(&self)
+    }
+    
+    @available(*, noasync)
+    mutating func unlock() {
+        pthread_mutex_unlock(&self)
+    }
+  }
+  ```
+ 
+  This attribute tells the compiler to raise a build error 
+  when the symbol is used in an asynchronous context. 
+  You can also use the `message` argument to provide additional information
+  about the symbol.
+
+  ```swift
+  @available(*, noasync, message: "Migrate locks to Swift concurrency.")
+  mutating func lock() {
+    pthread_mutex_lock(&self)
+  }
+  ```
+
+  If you can guarantee that your code uses a potentially unsafe symbol in a safe manner, 
+  You can wrap it in a synchronous function and call that function 
+  from an asynchronous context.
+
+  ```swift
+
+  // Provide a synchronous wrapper around methods with a noasync declaration.
+  extension pthread_mutex_t {
+    mutating func withLock(_ op: () -> ()) {
+      self.lock()
+      op()
+      self.unlock()
+    }
+  }
+
+  func downloadAndStore(key: Int,
+                      dataStore: MyKeyedStorage,
+                      dataLock: inout pthread_mutex_t) async {
+    // Safely call the wrapper in an asynchronous context.
+    dataLock.withLock {
+      dataStore[key] = downloadContent()
+    }
+  }
+  ```
+
+  You can use noasync availability with most declarations; 
+  however, you can't use it when declaring destructors.
+  The system must be able to call destructors from any context.
```

Comment:
I think that this bit could be a bit more consistent with the [deinitialization](https://docs.swift.org/swift-book/documentation/the-swift-programming-language/deinitialization) section. Instead of destructors, deinitializers is probably better.

Something like, "Swift but be able to call the deinitializer for a class from any context, regardless of whether it is synchronous or asynchronous"?